pkg/addoninstance/addoninstance.go

```diff
@@ -0,0 +1,44 @@
+package addoninstance
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+)
+
+// SetCondition sets a certain condition on an AddonInstance corresponding to the provided Addon
+// this function will be used by our tenants to report a heartbeat
+func SetCondition(ctx context.Context, kubeClient client.Client, condition metav1.Condition, addon addonsv1alpha1.Addon, log logr.Logger) error {
```

Comment:
Thanks for the suggestions @ashishmax31 :)
Regarding your first point, Yep, I thought about that as well but then, I realised that by making the `SetCondition` function accept `addonsv1alpha1.Addon` we're going to save 1 API call (to GET the addon) in certain situations. (instead of performing another unnecessary GET call to Kubernetes API for GETting the `addonsv1alpha1.Addon` object from Kubernetes API)

Let me try to explain

Say, for example, let's consider the "cluster-logging-operator" addon. And say, inside that addon's/operator's source code, there are 3 places/files where this `SetConditions` function is being called. And say, in 2 out of those 3 places/sections of its code, already the `addonsv1alpha1.Addon` object corresponding to the cluster-logging-operator is already available in the scope where `SetCondition` function is being called. Hence, in those 2 places, the developer could directly feed the `addonsv1alpha1.Addon` into the `SetConditions` function call. If `SetConditions` function were to accept `types.NamespacedName` as an argument, then, the developer would have to parse the Name/Namespace of the addon corresponding to cluster-logging-operator and feed it; and then internally behind the scenes, the `SetConditions` function would use that incoming `types.NamespacedName` to make an API call with Kubernetes client to GET that addon again which was available in the first place on developer's side/scope where SetConditions function was called (hence, an unnecessary extra API call to GET the addon would be performed).

Now, this was the best case scenario. But in the worst case, there's isn't any added concern. And it still is ensured that only 1 API call to GET the addon gets performed (on the side of the developer calling the `SetCondition` function).

Let me explain it better.

Say, In another situation (the 1 out of the 3 places where SetConditions function is being called), say there's a place/section of code where `SetCondition` function is called by the developer and in that scope, the `addonsv1alpha.Addon` object corresponding to cluster-logging-operator isn't present. Now, considering that the `SetCondition` function expects `addonsv1alpha.Addon` object (of cluster-logging-operator) as an argument, the developer can simply do a GET Api call with the Kubernetes client (r.Get) to get the `addonsv1alpha.Addon` object corresponding to their addon (cluster-logging-operator) and then, feed that received `addonsv1alpha.Addon` object as an argument to the `SetCondition` function. 
Why do I call this the worst case scenario? 
Because, In this situation, the developer would have to write 2-3 extra lines of code to GET the `addonsv1alpha.Addon` object but even in this situation, only 1 API call (of GETing the addon object) is being made to Kubernetes API in this whole lifecycle of the developer calling the `SetCondition` function. 

I know it sounds a bit overthought of but I am a stickler for runtime complexity 😛 

Do let me know if this sounds a lil convoluted and confusing, I'll try my best to explain in a more intuitive manner :)

Regarding your point around the possibility of getting rid of the `parseTargetNamespaceFromAddon` function. I am sorry but I don't understand how we could get rid of it in any situation (with `types.NamespacedName` arg or `addonsv1alpha1.Addon` arg). Because the purpose of `parseTargetNamespaceFromAddon` function is to fetch the `spec.install.olmOwnNamespace.namespace` or `spec.install.olmAllNamespaces.namespace` which can only be fetched from a `addonsv1alpha1.Addon` object which would be required to be processed in any situation (and this targetNamespace would get to be used later to install the `AddonInstance` resource in it)
So, I don't see how making use of `types.NamespacedName` correlates to getting rid of `parseTargetNamespaceFromAddon`. Do let me know if I am missing something :)

Cheers!

Reply:
@yashvardhan-kukreja  thanks for the explanation. I understand your concern, but why would an addon-developer already have an instance `addonsv1alpha.Addon` in their code, afaik this pkg would be the only thing they use inside their code.
But you're concern is valid, if they call this fn too many times within the heartbeatdeadline duration, its kind-of putting unnecessary load on the api-server w/o any real benefits.
Can we change this fn to something like:
```
registerCondition(ctx context.Context, kubeClient client.Client, log logr.Logger, addonName types.NamespacedName, conditionChan <- chan metav1.Condition) error
```
where we then in a go-routine periodically check the chan for status info and update the addon instance?
and addon-devs are expected to write status conditions to the channel from their side.
Wdyt?

Reply:
>  but why would an addon-developer already have an instance addonsv1alpha.Addon in their code, afaik this pkg would be the only thing they use inside their code.

So, they wouldn't directly need to do that in the first place but there might be certain places in the source code of the addon where in the same scope/section of code, `SetCondition` function might be getting called multiple times. In those situations, it would be efficient to GET the  `addonsv1alpha.Addon` object just once and feed the same object to every `SetCondition` call made in the same scope, rather than feeding ` types.NamespacedName`  to every `SetCondition` call made in that same scope, causing each `SetCondiiton` function call to make a GET request to Kubernetes API.

> Can we change this fn to something like:

Regarding this suggestion of yours around the `registerConditionChan`, I really love this approach. It's neat and efficient. But there's just only one tiny potential concern with that. 

So, my concern in this situation is that the addon devs would have to refactor their Addon reconciler's `struct` to contain the conditionChan as an attribute as well, and accordingly refactor the main.go file as well to register the Reconciler struct to the manager accordingly.

Let me show that.
Say, their addon is `cluster-logging-operator`
and the existing reconciler struct of it is
```go
type ClusterLoggingOperatorReconciler struct {
  client.Client
  Scheme *runtime.Scheme
}
```
They would have to edit it to
```go
type ClusterLoggingOperatorReconciler struct {
  client.Client
  Scheme *runtime.Scheme
  AddonInstanceConditionChan chan<- metav1.Condition
}
```
And in their main.go file, they would have to edit it this way:
```go
package main

import (
...
addoninstanceapi "github.com/mt-sre/addon-instance-sdk"
...
)
...
...
func main () {
...
...
  conditionChan , _ := addoninstanceapi. RegisterConditionChan(ctx, mgr.GetClient(), log, types.NamespacedName{Name: "addon-instance", Namespace: "namespace-of-the-addon"});

  // already existed section (except one line)
	if err = (&controllers.PostgresWriterReconciler{
		Client:           mgr.GetClient(),
		Scheme:           mgr.GetScheme(),
		AddonInstanceConditionChan: conditionChan,  //newly added line
	}).SetupWithManager(mgr); err != nil {
		setupLog.Error(err, "unable to create controller", "controller", "ClusterLoggingOperator")
		os.Exit(1)
	}
	//+kubebuilder:scaffold:builder
...
...
}
```

Now, **assuming that** essentially anywhere across the addon's codebase, the `ClusterLoggingOperatorReconciler` object would be accessible, because majority of the methods would be the methods of `ClusterLoggingOperatorReconciler`. This would mean that anywhere across the codebase the condition could be set just in one-line of code via passing the metav1.condition to the `AddonInstanceConditionChan` of the `ClusterLoggingOperatorReconciler` object in the scope.
Let me demonstrate that as well:

For example, `Reconcile()` method.

```go
func (r *ClusterLogginOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
...
...
// if I want to report a condition here
r.AddonInstanceConditionChan <- metav1.Condition{Type: "Healthy",  Status: "True", Reason: "yay!", Messgae: "yay!"}

// or if I want another condition to be here
r.AddonInstanceConditionChan <- metav1.Condition{Type: "Healthy",  Status: "False", Reason: "Ouch!", Messgae: "Ouch!"}

}
```
This is because I am assuming that in any place of their addon's codebase, wherever they would to register/set condition of the `AddonInstance`, that place would have access to the `ClusterLoggingOperatorReconciler` object.

And finally, in the `main.go` file of the Addon's codebase, a go routine can be written to receive any `metav1.Condition` from the `AddonInstanceConditionChan` of that addon in almost real-time and whenever it would receive a metav1.Condition, it will feed that to `addoninstanceapi.SetCondition` function and register that condition.


---

Another refactor-level I am afraid of is that this approach might bother the addon developer further to edit/refactor the sections/scopes of their addon's code where the Reconciler object is not accessible. They would have to refactor those portions of code to accept the `AddonInstanceConditionChan` as an argument as well..
For example, say they want to Set a condition in an existing `Foo` function in the addon's code.
They would have to edit their function signature from
```go
func Foo (arg1, arg2, arg3) {

// here, they want to set the condition

}
```
to 

```go
func Foo(arg1, arg2, arg3, addonInstanceConditionChan chan<- metav1.Condition) {
  // here, they want to set the condition
  addonInstanceConditionChan <- metav1.Condition{Type: .. , Reason: ..., Status: ...., }
}
```
And this would further mean to refactor/edit all the places where the above `Foo` function is getting called in the addon's code.

---

So, these are the concerns which are tingling in my head. I am still not that much concerned about the first refactor-level concern which I mentioned around refactoring the Reconciler `struct` because that's barely a one-time and 1-2 line edit (just adding  a new attribute in the Reconciler struct and populating it once in main.go). But I am more concerned about the second refactor-level concern I mentioned around editing the function signatures of those functions in addon's code where conditions are supposed to be set but the Reconciler object isn't reachable. (potentially substantial refactor would be required).

But honestly, I love this approach and will implement it as well to see how it works out. And other team members can put in their thoughts as well around how it looks :)

Thanks!

Reply:
Yeah like you said the 1st point wouldn't be a prob, we just tell them its the way to integrate with our SDK. But the second one would be a potential prob, but that depends on how their code is structured.